src/ipc/handler.c

```diff
@@ -815,6 +817,147 @@ static int ipc_glb_gdb_debug(uint32_t header)
 
 }
 
+#if CONFIG_PROBE
+static inline int ipc_probe_init(uint32_t header)
+{
+	struct sof_ipc_probe_dma_add_params *params = _ipc->comp_data;
+	int dma_set_up = (params->hdr.size - sizeof(*params)) /
+			 sizeof(struct probe_dma);
+
+	tracev_ipc("ipc_probe_init()");
+
+	return probe_init(dma_set_up ? params->probe_dma : NULL);
+}
+
+static inline int ipc_probe_deinit(uint32_t header)
+{
+	tracev_ipc("ipc_probe_deinit()");
+
+	return probe_deinit();
+}
+
+static inline int ipc_probe_dma_add(uint32_t header)
+{
+	struct sof_ipc_probe_dma_add_params *params = _ipc->comp_data;
+	uint32_t dmas_count = (params->hdr.size - sizeof(*params)) /
```

Comment:
what happens if hdr.size is too small, that would wrap-arount with the unsigned int representation?

Reply:
Then hell breaks loose and we're running over the memory until something bad happens, possibly creating some probes unintentionally.
As you've already noticed, we have a strong dependency on valid `hdr.size` being passed from kernel, and if it fails, there aren't really many ways to cope with that. I could add a check here, but it wouldn't be bulletproof anyway.

Reply:
you have a max in the Kconfig so you should at least check the values are in the expected range.

Reply:
If such check fails, then we discard whole IPC command, correct?
Without such check, what we currently do is to parse as many as possible, until limit is reached or until we encounter invalid data. Only then we report `-EINVAL`, but all actions completed up to that point are still considered valid and operational.

Reply:
I would err on the side of rejecting an entire IPC if there is an error, be it at the start (invalid size) or later during parsing.
Partial commits are usually not a great idea. From the kernel perspective if the IPC fails then it's not considered to have done anything (states don't change, etc).

Reply:
@plbossart AFAIK, in case anything goes wrong with these probes IPCs, a INFO call is made to check the actual state of FW and corrections are made from that point. Possibly even after a successful call, kernel still monitors if probes are in the right state on FW side.
Also, rewinding state on FW side to point before IPC processing begun, would probably be messy and in the end, the negotiation of state that I mentioned seems to be the optimal solution in the long run.
@crojewsk could You correct me if I'm wrong on this subject?

Reply:
if anything goes sideways in the IPC, the kernel will not try and figure out what parts went through. It's just massively complicated to do so, we'll throw an error and stop. Error resilience is usually not worth the effort, it's just simpler to fail big and early.

Reply:
I think that we could drag @crojewsk into this conversation, and properly discuss this, as we move onto implementation details. For now, let's focus on the interface, so that we could move on, please. Surely, these implementation details are important, and I don't want to neglect them, but I would really like to see this merged this year.

Reply:
@plbossart checking something like hdr size should not be a part of any module like probes, if we want to check something like this we should do this at the beginning of processing IPC, 
In ipc_cmd() we already have one check for "ipc: invalid IPC header", it can be extended if needed.

Reply:
I agree here with @abonislawski suggestion. If we want to extend the scope of IPC header validation then this should be done on IPC dispatcher level and not within each single module.

Reply:
@plbossart @lgirdwood Added `sof_ipc_flexible_array_hdr` as secondary header to the payload. Lengths of the arrays are now explicitly defined in every probe IPC command.